src/chips/sam4l/flashcalw.rs

```diff
@@ -0,0 +1,1106 @@
+/* chips::sam4l::flashcalw -- Implementation of a flash controller.
+ *
+ * This implementation of the flash controller for at sam4l flash controller
+ * uses interrupts to handle main tasks of a flash -- write, reads, and erases.
+ * If modifying this file, you should check whether the flash commands (issued
+ * via issue_command) generates and interrupt and design a higher level function
+ * based off of that.
+ *
+ * Although the datasheet says that when the FRDY interrupt is on, a interrupt will
+ * be generated after a command is complete, it doesn't appear to occur for some
+ * commands.
+ *
+ * A clean interface for reading from flash, writing pages and erasing pages are
+ * defined below and should be used to handle the complexity of these tasks.
+ * 
+ * The driver should be configured() before use, and a Client should be set to
+ * enable a callback after a command is completed.
+ * Author:  Kevin Baichoo <kbaichoo@cs.stanford.edu>
+ * Date: 7/27/16
+ */
+
+use helpers::*;
+use core::{mem, slice};
+use core::cell::Cell;
+use common::take_cell::TakeCell;
+use pm;
+use nvic;
+
+//  These are the registers of the PicoCache -- a cache sole for the flash.
+#[allow(dead_code)]
+struct Picocache_Registers {
+    reserved_1:                             [u8;0x8],
+    picocache_control:                      usize,
```

Comment:
the `picocache_` prefixes are not necessary, since these are already fields of the picocache struct